### ISSUE TITLE
undeterministic behavior with cuDNN and equal random seed

### ISSUE THREAD

**Issue by MartinHahner** [NVIDIA/caffe#432]
### When I build caffe with cuDNN (Feb2017 release) I cannot reproduce training results.

Screenshot from DIGITS:
![2017-10-13_15 26 20](https://user-images.githubusercontent.com/14181188/31610034-d79d54c4-b276-11e7-97ad-ca17c549440a.jpg)

All experiments named "_AlexNet Xp_" have the same random seed:
![2017-10-16_13 39 05](https://user-images.githubusercontent.com/14181188/31610194-754b504a-b277-11e7-847a-4c52aa3fb342.jpg)

This random seed also finds its way into the solver.prototxt:

> 
> _test_iter: 391
> test_interval: 293
> base_lr: 0.01
> display: 36
> max_iter: 8790
> lr_policy: "step"
> gamma: 0.1
> momentum: 0.9
> weight_decay: 0.0001
> stepsize: 2901
> snapshot: 293
> snapshot_prefix: "snapshot"
> solver_mode: GPU
> **random_seed: 1**
> net: "train_val.prototxt"
> solver_type: SGD_

So the issue is not in DIGITS, 
it also happens when I invoke caffe with the mentioned solver.prototxt directly:
_./caffe train --solver=/solver.prototxt --gpu=0_

**training run 1**
> Test net output 0: accuracy = **0.631634**
> Test net output 1: loss = 1.06598 (* 1 = 1.06598 loss)

**training run 2**
> Test net output 0: accuracy = **0.559942**
> Test net output 1: loss = 1.23086 (* 1 = 1.23086 loss)

**training run 3**
> Test net output 0: accuracy = **0.564019**
> Test net output 1: loss = 1.23072 (* 1 = 1.23072 loss)

### However, when I  build caffe without cuDNN I get deterministic behavior as expected:

**training run 1**
> Test net output 0: accuracy = **0.560822**
> Test net output 1: loss = 1.22098 (* 1 = 1.22098 loss)

**training run 2**
> Test net output 0: accuracy = **0.560822**
> Test net output 1: loss = 1.22098 (* 1 = 1.22098 loss)

**training run 3**
> Test net output 0: accuracy = **0.560822**
> Test net output 1: loss = 1.22098 (* 1 = 1.22098 loss)

I am using caffe v0.16.4-2-gcdb3d9a5.

**Comment by drnikolaev** [comment]
Hi @MartinHahner88 thank you. Not all cuDNN implementations of Convolution and Pooling algorithms are deterministic.

**Comment by MartinHahner** [comment]
Thank you @drnikolaev for your reply. 
Do you have a reference to more details about the non-deterministic behaviour of cuDNN 
or do you know more details about it by yourself? For example why?

Is there a way to force determinism using cuDNN?

I found significant speed boosts using cuDNN. 
I did not measure it, but I would say I roughly had 5x faster training times using cuDNN.

**Comment by drnikolaev** [comment]
The reason why is simple: performance. If you need pure determinism you have to use deterministic algorithms which might be slower. For that you need to tweak your model but I'm not sure that DIGITS supports that.
